### PR TITLE
Fix Specifying Initial States of RNN Layers

### PR DESCRIPTION
**Pull request by Joshua-Chin**
I have begun to review the changes made in da21c15. There appears to a number of issues

- If `initial_states` is a list of tensors, it will not have the attribute `_keras_history`, and will be treated as a numerical value.
- If `initial_states` is passed as a keyword argument to `call`, it will be ignored / overwritten.
- `state_spec` may not be defined by the time it is used (`state_spec` is defined in `build`).
- In `reset_states`, there is a check if `input_spec` is not `None`. However, `input_spec` is never `None`, because it is defined in `__init__` to be `InputSpec(ndim=3)`.
- There is an inconsistency between using the singular and plural of `initial_state` and `initial_states`.
- the signature for `reset_states(state_values)` is inconsistent with the signature for  `set_weights(weights)`
- `test_specify_initial_states` does not check if `initial_states` is part of the computational graph.

This commit does the following to fix those issues:

- If `initial_states` is passed to `__call__` it is always treated as a tensor / list of tensors. It a user wants to specify the state numerically, they should pass a numpy array to `reset_states`.
- The layer will be build in `__call__` before using `state_spec`.
- The extraneous check of `input_spec` is removed.
- The plural form `initial_states` is used throughout the code and documentation.
- The signature for `reset_states(state_values)` is changed to `reset_states(states)`.
- `test_specify_states` explicitly checks if `initial_states` is added to the computational graph.

#5738

### CODE REVIEW

**Comment by fchollet** [comment]
> If initial_states is a list of tensors, it will not have the attribute _keras_history, and will be treated as a numerical value.

The proper behavior in that case would to check if all tensors in the list are Keras tensors. If it's mixed, raise an exception. If they're all Keras tensors, add them to the inputs.

I am confused by what you are referring to as "numerical values". There are no numerical values involved (e.g. Numpy tensors), only symbolic tensors (e.g. TF tensors). Can you clarify?

**Comment by fchollet** [comment]
> The plural form initial_states is used throughout the code and documentation.

In the API we should use `initial_state` everywhere, for API consistency with TF. In the code we can use whatever.

**Comment by Joshua-Chin** [comment]
> The proper behavior in that case would to check if all tensors in the list are Keras tensors. If it's mixed, raise an exception. If they're all Keras tensors, add them to the inputs. Else, use initial_state.

It is unclear to me why we are checking if they are Keras tensors in the first place. If `initial_state` does not depend on the input, it must be a constant and should be specified in `reset_states`. In addition, it may cause issues for people using Keras as a layer library over TF / Theano.

>  I am confused by what you are referring to as "numerical values". There are no numerical values       involved (e.g. Numpy tensors), only symbolic tensors (e.g. TF tensors). Can you clarify?

My apologies, I misspoke. In the current code, if `initial_state` is a list of tensors, it won't be treated as numerical values.
However, in the current code, the value of `initial_state` will be ignored. The code block from 263-269 in `Recurrent.call` will always overwrite the value of `initial_state` keyword argument.

> In the API we should use initial_state everywhere, for API consistency with TF. In the code we can use whatever.

I am fine with using initial_state everywhere in the API. However, I feel that we should change the code in `recurrent.py` to always use initial_state. Switching between `initial_state` and `initial_states` adds an unnecessary overhead.

**Comment by fchollet** [comment]
> If initial_state does not depend on the input, it must be a constant and should be specified in reset_states. In addition, it may cause issues for people using Keras as a layer library over TF / Theano.

A non-Keras tensor set as initial state will generally not be a constant. It will simply be a non-Keras tensor, dependent or not on the underlying model's inputs.

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -228,50 +233,61 @@ def __call__(self, inputs, initial_state=None, **kwargs):
...
-                # Perform the call
-                output = super(Recurrent, self).__call__(inputs, **kwargs)
-
-                # Restore original input spec
-                self.input_spec = input_spec
-                return output
-            else:
-                kwargs['initial_state'] = initial_state
+
+            # We need to build the layer so that state_spec exists.
```
Comment (fchollet) [inline keras/layers/recurrent.py]: All of this should be delegated to the parent's `__call__`.

File: keras/layers/recurrent.py [hunk 2]
```diff
@@ -228,50 +233,61 @@ def __call__(self, inputs, initial_state=None, **kwargs):
...
+            input_spec = self.input_spec
+            state_spec = self.state_spec
+            if not isinstance(state_spec, list):
+                state_spec = [state_spec]
+            self.input_spec = [input_spec] + state_spec
+
+            # Compute the full inputs, including state
+            if not isinstance(initial_state, (list, tuple)):
+                initial_state = [initial_state]
+            inputs = [inputs] + list(initial_state)
```
Comment (fchollet) [inline keras/layers/recurrent.py]: There is no longer any check that the initial states are Keras tensors, which will cause the model construction to fail when using non-Keras tensors.

File: keras/layers/recurrent.py (hunk 1 above)
Comment (Joshua-Chin) [inline keras/layers/recurrent.py]: Currently, `state_spec` is defined in `Recurrent.build`. When, `initial_state` is passed, we need to build the layer so that we can use `state_spec`.

File: keras/layers/recurrent.py (hunk 2 above)
Comment (Joshua-Chin) [inline keras/layers/recurrent.py]: For every other type of layer, model construction fail when using non-Keras tensors. Why are we making a special exception for the initial states of RNNs?

Comment (fchollet) [inline keras/layers/recurrent.py]: We're not making an exception, RNN layers will behave like every other layer. We're just building an API. The API is that `RNN(inputs, initial_state=x)` should work as a way to set initial state tensors, independently of the value of `x` (Keras tensor or not). It's actually very simple to set up, via a switch between inputs and layer keyword arguments.

**Comment by Joshua-Chin** [comment]
The latest commit moves the definition of `state_spec` to `__init__`, so we don't have redundant code in `Recurrent.__call__` to build the layer. In addition, it properly handles the case where `initial_state` may be a list of Keras tensors.

**Comment by fchollet** [comment]
Any volunteers to review this PR?

**Comment by smodlich** [comment]
Any updates on this? Setting the initial state seems to be an important component of any viable Seq2Seq-Model.

**Comment by farizrahman4u** [comment]
Looks good. Please merge if no other issues.

**Comment by AMabona** [comment]
Just a heads up, this breaks because of masking in a seq2seq model with the tensorflow backend. This is because `_collect_previous_masks` returns a list with the decoder RNN mask and `None` (since initial state is the encoder's final hidden state which has no mask). This is ultimately passed to the tensorflow backend RNN function which throws an error because it expects a tensor as the mask, but gets a list.

Something similar will happen when you pass any tf tensor as the initial state (e.g. image captioning, RNN VAE). It will happen whenever either the initial state or the RNN has a mask.

In gratuituous detail:
1. In `recurrent.py`, the `initial_state` is added to `inputs` and `topology.__call__` is called on it.
2. In `topology.__call__`,  `previous_mask` is computed, it's a list containing the decoder RNN mask and `initial_state`'s mask. `Recurrent.call` is called with this mask.
3. `Recurrent.call` calls `K.rnn` with the mask which falls over because it's a list.

**Comment by farizrahman4u** [comment]
@fchollet I think it would be nice to have Keras topology handle optional inputs. First step would be to make it such a way that input spec for layers with multiple inputs is not a list, but a single `InputSpec` object  (with additional attributes num_inputs, num_optional_inputs).

**Comment by farizrahman4u** [comment]
@Joshua-Chin Have sent you a pull request to fix masking issue mentioned by @AMabona.

**Comment by fchollet** [comment]
@farizrahman4u currently only this layer will require optional inputs, so an initial ad-hoc system is fine in this case. Later we can use what we learned from this ad-hoc implementation to write a more general system that can apply to all layers.

**Comment by fchollet** [comment]
Could someone please review this PR?

**Comment by farizrahman4u** [comment]
@fchollet Should reset_states be renamed to reset_state (with backward compatibility) for consistency?

**Comment by Joshua-Chin** [comment]
@farizrahman4u @fchollet What's the status of the review?

**Comment by farizrahman4u** [comment]
I think we are good.. fix the typo though.

File: keras/layers/recurrent.py [hunk 3]
```diff
@@ -227,51 +234,62 @@ def __call__(self, inputs, initial_state=None, **kwargs):
...
+
+        is_keras_tensor = hasattr(initial_state[0], '_keras_history')
+        for tensor in initial_state:
+            if hasattr(tensor, '_keras_history') != is_keras_tensor:
+                raise ValueError('The initial state of an RNN layer cannot be'
+                                 ' specified with a mix of Keras tensors and'
+                                 ' non-Keras tensors')
+
+        if is_keras_tensor:
+
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Unnecessary blank line
